### PR TITLE
ReturnToPrevious: Create the feature toggle, the wrapper and the button group

### LINKED ISSUE #79881: Create the `ReturnToPrevious` feature toggle, wrapper and button group
Reference: [POC: Return to previous floating button logic](https://github.com/grafana/grafana/pull/79688)

- [x] Create the feature toggle: [Feature toggle guide](https://github.com/grafana/grafana/blob/3bcde852acf91739a41014fe3179603540d89247/contribute/feature-toggles.md#L1)
https://github.com/grafana/grafana/pull/80831
- [ ] ~~Create the ReturnToPreviousWrapper~~: Finally not needed
- ~~If the 'returnToPreivous' feature toggle is on, show the `ReturnToPreviousWrapper`. This condition was added in the `AppWrapper`. This component, the `ReturnToPreviousWrapper`, evaluates if the URL params exist. An example can be found in the reference.~~
- ~~If the user goes back to the URL that matches the one in the params it will reset the functionality and the button group will disappear.~~
- [x] Create the ReturnToPrevious ButtonGroup:  
https://github.com/grafana/grafana/pull/80878
- As in the previous task, there is an example in the reference POC. 
- The first button is the one that contains the URL to go back to and the title of the page to return.
- The title should have a maximum of characters shown set to 40. Large titles will be shortened and will appear with an ellipsis.
- The second button is also a `Button` component, even just showing a close icon. It will make the button group disappear and get rid of the URL params. 
- ButtonGroup appearance: primary + outlined

- [x] Add logic to show the new component in `AppChrome` behind the new toggle: 
https://github.com/grafana/grafana/pull/81035
___

**Important update:** To keep the characteristics we agreed the 'ReturnToPrevious' should have, we finally changed the use of `URL params` for the use of `sessionStorage`, as it allows the button to be shown even when the user continues browsing and, therefore, changing the URL.

### PR DESCRIPTION
**Pull request by eledobleefe**
**What is this feature?**

This feature brings an option to let the user go back to the previous context (that triggered the functionality).

**Why do we need this feature?**

To offer an option to return to the previous context.

**Who is this feature for?**

Everybody

**Which issue(s) does this PR fix?**:

Fixes #79881 

**Special notes for your reviewer:**
To test this, you can modify the [RuleDetailsActionButtons.tsx](https://github.com/grafana/grafana/blob/6c609051a0af2fad8f9f6918b93ca9397c147ddb/public/app/features/alerting/unified/components/rules/RuleDetailsActionButtons.tsx#L144), adding this to the 'Go to dashboard' button:
```jsx
onClick={() => {
            setReturnToPrevious({ title: rule.name, href: locationService.getLocation().pathname });
            locationService.push(`/d/${encodeURIComponent(dashboardUID)}`);
          }}
```
Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  const params = useQueryParams()[0];
+  const styles = useStyles2(getStyles);
+  const theme = useTheme2();
+
+  const [paramsExist, setParamsExist] = React.useState(params?.__returnToTitle && params?.__returnToUrl);
+  // Don't show the button whether if there is no params or the URL param matches the current URL
+  const showReturnToPrevious: boolean = paramsExist && location.pathname !== params.__returnToUrl ? true : false;
+
+  // Only show the button on large screens
+  const isLargeScreen = window.innerWidth >= theme.breakpoints.values.xl;
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx:20]: Does this mean we are not too worried about window resizing? If we want to consider window resizing we can put this in a media query instead?

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import { ReturnToPrevious } from './ReturnToPrevious';
+
+export const ReturnToPreviousWrapper = () => {
+  const params = useQueryParams()[0];
+  const styles = useStyles2(getStyles);
+  const theme = useTheme2();
+
+  const [paramsExist, setParamsExist] = React.useState(params?.__returnToTitle && params?.__returnToUrl);
+  // Don't show the button whether if there is no params or the URL param matches the current URL
+  const showReturnToPrevious: boolean = paramsExist && location.pathname !== params.__returnToUrl ? true : false;
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx:17]: I think we can merge this logic, something like 

```
const [showReturnToPrevious, setShowReturnToPrevious] = React.useState(params?.__returnToTitle && params?.__returnToUrl && location.pathname !== params.__returnToUrl);
```

and then later in the useEffect also just set this directly. we could create a small helper

```
const shouldShow = (params) => params?.__returnToTitle && params?.__returnToUrl && location.pathname !== params.__returnToUrl
```

and use it in both. Wdyt?

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const styles = useStyles2(getStyles);
+  const [, setParams] = useQueryParams();
+
+  const handleOnClick = () => {
+    href && locationService.push(href.toString());
+  };
+  const closeButton = () => {
+    setParams({ __returnToTitle: null, __returnToUrl: null });
+    const currentLocation = locationService.getLocation();
+    locationService.push(currentLocation);
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx:25]: This creates a strange pattern to me, when I press the 'x' and then press the browser back button it keeps the same URL, one more back and it readds the params to the URL. I think we need to investigate this and maybe use `locationService.replace`, but I would double check what is the intended UX here.

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  };
+
+  const shortenTitle = (title: UrlQueryValue) => {
+    const titleLength = 37;
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx:29]:
```suggestion
    const maxTitleLength = 37;
```

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    return title && title.toString().length > titleLength ? title.toString().substring(0, titleLength) + '...' : title;
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx:30]: If we already have text overflow ellipsis do we need this calculation still?

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      setParamsExist(true);
+    } else {
+      setParamsExist(false);
+    }
+  }, [params]);
+
+  return showReturnToPrevious && isLargeScreen ? (
+    <div className={styles.wrapper}>
+      <ReturnToPrevious href={params.__returnToUrl} title={params.__returnToTitle}>
+        {params.__returnToTitle}
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPreviousWrapper.tsx:33]: Why do we pass title twice as `children` and as `title` prop?

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      <Button
+        icon="angle-left"
+        size="sm"
+        variant="primary"
+        fill="outline"
+        onClick={handleOnClick}
+        title={title?.toString()}
+        className={styles.returnToPrevious}
+      >
+        Back to {shortenTitle(children)}
```
Comment (JoaoSilvaGrafana) [inline public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx:44]: I know this is draft but just a reminder to translate this and the aria-label 😄

**Review by JoaoSilvaGrafana — COMMENTED** [review]
This looks really good 👏  Been playing around with it and left some questions

**Review by ashharrison90 — COMMENTED** [review]
an aside, but i really think this should have been 3 separate PRs

1. new feature toggle
1. new component
1. the 'return to previous' logic to show the new component in `AppChrome` behind the new toggle

not only is everything easier to review then, but by doing it this way it forces you to think about the component in isolation. as it is, the `ReturnToPrevious` component is hyper specific to this use case, when instead it could have been named something generic like `DismissableButton` with `label`, `onClick` and `onClose` properties.

**Comment by eledobleefe** [comment]
> an aside, but i really think this should have been 3 separate PRs
> 
>     1. new feature toggle
> 
>     2. new component
> 
>     3. the 'return to previous' logic to show the new component in `AppChrome` behind the new toggle
> 
> 
> not only is everything easier to review then, but by doing it this way it forces you to think about the component in isolation. as it is, the `ReturnToPrevious` component is hyper specific to this use case, when instead it could have been named something generic like `DismissableButton` with `label`, `onClick` and `onClose` properties.

Hi @ashharrison90,
First of all, thanks for your comments.
Regarding this, I do not mind splitting it now. To be honest, I also thought about it, but I didn't do it in the end 🤷 
About the isolation part, should we always aim to create generic components first? Even for a specific functionality?

File: public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  return (
+    <ButtonGroup className={styles.wrapper}>
+      <Button
+        icon="angle-left"
+        size="sm"
+        variant="primary"
+        fill="outline"
+        onClick={handleOnClick}
+        title={title?.toString()}
```
[the remark this answers is not in the document]
Comment (eledobleefe) [reply public/app/core/components/AppChrome/ReturnToPrevious/ReturnToPrevious.tsx]: Good catch! It should be needed once and then I forgot it was there 😓

File: public/app/core/components/AppChrome/AppChromeService.tsx [hunk 9]
```diff
@@ -108,6 +120,10 @@ export class AppChromeService {
     return useObservable(this.state, this.state.getValue());
   }
 
+  // public setReturnToPrevious(returnToPrevious: AppChromeState['returnToPrevious']) {
+  //   this.update({ returnToPrevious });
+  // }
+
```
Comment (L-M-K-B) [inline public/app/core/components/AppChrome/AppChromeService.tsx:126]: Do you still need this for something?

**Comment by eledobleefe** [comment]
Closing this PR since this was split into three smaller PRs that already got merged.

**Pull request closed without merging**